File: src/python-app/src/services/group_service.py
"""Service for managing host groups."""

from dataclasses import dataclass

from src.domain.entities import RecordType
from src.domain.exceptions import FirewallOperationException, ResourceAlreadyExistsException
from src.infrastructure.firewall_client import FirewallClient


@dataclass
class GroupConfiguration:
    """Configuration for host groups."""
    
    base_name: str
    
    @property
    def fqdn_group_name(self) -> str:
        """Get FQDN group name."""
        return f"{self.base_name}_FQDNHostGroup"
    
    @property
    def ip_group_name(self) -> str:
        """Get IP group name."""
        return f"{self.base_name}_IPHostGroup"


class HostGroupService:
    """
    Service for host group operations.
    
    Handles creation and management of FQDN and IP host groups.
    """
    
    def __init__(self, firewall_client: FirewallClient, config: GroupConfiguration):
        """
        Initialize service.
        
        Args:
            firewall_client: Client for firewall operations
            config: Group configuration
        """
        self._client = firewall_client
        self._config = config
    
    @property
    def fqdn_group(self) -> str:
        """Get FQDN group name."""
        return self._config.fqdn_group_name
    
    @property
    def ip_group(self) -> str:
        """Get IP group name."""
        return self._config.ip_group_name
    
    def create_groups(self) -> dict[str, bool]:
        """
        Create both FQDN and IP host groups.
        
        Returns:
            Dictionary with group names and creation status
        """
        results = {}
        
        # Create FQDN group
        try:
            self._client.create_fqdn_group(self.fqdn_group)
            results[self.fqdn_group] = True
        except ResourceAlreadyExistsException:
            results[self.fqdn_group] = False
        except FirewallOperationException as e:
            # Check if it's an "already exists" error from the error message
            error_msg = str(e).lower()
            if "already exists" in error_msg or "same name" in error_msg:
                results[self.fqdn_group] = False
            else:
                raise  # Re-raise if it's a different error
        
        # Create IP group
        try:
            self._client.create_ip_group(self.ip_group)
            results[self.ip_group] = True
        except ResourceAlreadyExistsException:
            results[self.ip_group] = False
        except FirewallOperationException as e:
            # Check if it's an "already exists" error from the error message
            error_msg = str(e).lower()
            if "already exists" in error_msg or "same name" in error_msg:
                results[self.ip_group] = False
            else:
                raise  # Re-raise if it's a different error
        
        return results
    
    def get_group_for_record_type(self, record_type: RecordType) -> str:
        """
        Get appropriate group name for a record type.
        
        Args:
            record_type: Type of network record
            
        Returns:
            Group name
            
        Raises:
            ValueError: If record type doesn't map to a group
        """
        if record_type == RecordType.FQDN:
            return self.fqdn_group
        
        if record_type in (RecordType.IP_ADDRESS, RecordType.NETWORK_CIDR):
            return self.ip_group
        
        raise ValueError(f"No group mapping for record type: {record_type}")

File: src/rust-app/src/services/group_service.rs
//! Host group management service. Ports `src/services/group_service.py`.

use crate::domain::entities::RecordType;
use crate::domain::errors::{AppError, Result};
use crate::infra::firewall_client::FirewallApi;

/// Configuration for host group names.
pub struct GroupConfiguration {
    pub base_name: String,
}

impl GroupConfiguration {
    pub fn new(base_name: impl Into<String>) -> Self {
        Self {
            base_name: base_name.into(),
        }
    }

    pub fn fqdn_group_name(&self) -> String {
        format!("{}_FQDNHostGroup", self.base_name)
    }

    pub fn ip_group_name(&self) -> String {
        format!("{}_IPHostGroup", self.base_name)
    }
}

/// Service for creating and resolving FQDN and IP host groups.
pub struct HostGroupService<'a> {
    client: &'a dyn FirewallApi,
    fqdn_group: String,
    ip_group: String,
}

impl<'a> HostGroupService<'a> {
    pub fn new(client: &'a dyn FirewallApi, config: GroupConfiguration) -> Self {
        Self {
            client,
            fqdn_group: config.fqdn_group_name(),
            ip_group: config.ip_group_name(),
        }
    }

    pub fn fqdn_group(&self) -> &str {
        &self.fqdn_group
    }

    pub fn ip_group(&self) -> &str {
        &self.ip_group
    }

    /// Create both host groups. Returns `(group_name, created)` pairs in
    /// creation order (`created = false` means the group already existed).
    pub fn create_groups(&self) -> Result<Vec<(String, bool)>> {
        let fqdn =
            self.create_one(self.client.create_fqdn_group(&self.fqdn_group), &self.fqdn_group)?;
        let ip = self.create_one(self.client.create_ip_group(&self.ip_group), &self.ip_group)?;
        Ok(vec![fqdn, ip])
    }

    /// Interpret a single group creation outcome, treating "already exists"
    /// errors as `created = false` and propagating anything else.
    fn create_one(&self, outcome: Result<()>, name: &str) -> Result<(String, bool)> {
        match outcome {
            Ok(()) => Ok((name.to_string(), true)),
            Err(AppError::ResourceAlreadyExists(_)) => Ok((name.to_string(), false)),
            Err(AppError::FirewallOperation {
                message,
                status_code,
            }) => {
                let lower = message.to_lowercase();
                if lower.contains("already exists") || lower.contains("same name") {
                    Ok((name.to_string(), false))
                } else {
                    Err(AppError::FirewallOperation {
                        message,
                        status_code,
                    })
                }
            }
            Err(other) => Err(other),
        }
    }

    /// Group name for a record type. FQDN → FQDN group; IP/CIDR → IP group.
    pub fn group_for_record_type(&self, record_type: RecordType) -> Result<&str> {
        match record_type {
            RecordType::Fqdn => Ok(&self.fqdn_group),
            RecordType::IpAddress | RecordType::NetworkCidr => Ok(&self.ip_group),
            RecordType::Invalid => Err(AppError::Validation(format!(
                "No group mapping for record type: {}",
                record_type.as_str()
            ))),
        }
    }
}
